### PR TITLE
Add field to expose entrypoint num cpus in rayjob

### LINKED ISSUE #1266: [Feature] [RayJob] Support `entrypoint_num_cpus`, `gpus`, `resources` arguments
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

https://docs.ray.io/en/latest/cluster/running-applications/job-submission/sdk.html#specifying-cpu-and-gpu-resources

The Ray Job API supports specifying resources, but these options are not exposed as fields in the KubeRay RayJob spec.  We should add them.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by shubhscoder**
## Why are these changes needed?

Adds the fields entrypoint_num_cpus, gpus, resources in Kuberay Rayjob spec. Ray Job API supports specification of these resources, but before this code change the Kuberay Job Spec did not expose these fields  

## Related issue number

Closes #1266

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by shubhscoder** [comment]
@architkulkarni I think the code changes are ready for review. I have tried to complete all the steps given in the development guide. Apologies if something is missing. Also I am not sure why the tests are failing, I don't seem to have access to the logs. Thanks in advance for the help!

File: ray-operator/controllers/ray/common/job.go [hunk 1]
```diff
@@ -108,6 +111,18 @@ func GetK8sJobCommand(rayJobInstance *rayv1alpha1.RayJob) ([]string, error) {
...
+	if entrypointNumCpus > 0 {
+		k8sJobCommand = append(k8sJobCommand, "--entrypoint_num_cpus", fmt.Sprintf("%f", entrypointNumCpus))
+	}
+
+	if entrypointNumGpus > 0 {
+		k8sJobCommand = append(k8sJobCommand, "--entrypoint_num_gpus", fmt.Sprintf("%f", entrypointNumGpus))
+	}
+
+	if len(entrypointResources) > 0 {
+		k8sJobCommand = append(k8sJobCommand, "--entrypoint_resources", entrypointResources)
```
Comment (architkulkarni) [inline ray-operator/controllers/ray/common/job.go:123]:
```suggestion
	if entrypointNumCpus > 0 {
		k8sJobCommand = append(k8sJobCommand, "--entrypoint-num-cpus", fmt.Sprintf("%f", entrypointNumCpus))
	}

	if entrypointNumGpus > 0 {
		k8sJobCommand = append(k8sJobCommand, "--entrypoint-num-gpus", fmt.Sprintf("%f", entrypointNumGpus))
	}

	if len(entrypointResources) > 0 {
		k8sJobCommand = append(k8sJobCommand, "--entrypoint-resources", entrypointResources)
```

**Review by architkulkarni — COMMENTED** [review]
Looks good to me, but there's a typo.  To catch this we would have needed an integration test with Ray. Can you add one? Here's one idea.

- Add a new sample YAML file which has the job specify some cpus, gpus, and resources.  I think there should be some way to also specify these logical resources in the RayCluster spec for the job (since we don't have physical GPUs, we don't want to autodetect the number of GPUs, but we can just define that the cluster has 4 GPUs or something.)
- In the entrypoint script, use `ray.available_resources()` and `ray.cluster_resources()` to ensure the expected number of resources are taken up by the currently running script.
- Add the name of the new YAML file to `test_sample_rayjob_yamls.py` so that it's tested in CI.

What do you think?

**Comment by architkulkarni** [comment]
Can you say more about "don't have access to the logs"? You should be able to run tests locally using `kind` and if logs are missing, it could be a bug or we might need to improve the development guide.

**Comment by shubhscoder** [comment]
> Looks good to me, but there's a typo. To catch this we would have needed an integration test with Ray. Can you add one? Here's one idea.
> 
> * Add a new sample YAML file which has the job specify some cpus, gpus, and resources.  I think there should be some way to also specify these logical resources in the RayCluster spec for the job (since we don't have physical GPUs, we don't want to autodetect the number of GPUs, but we can just define that the cluster has 4 GPUs or something.)
> * In the entrypoint script, use `ray.available_resources()` and `ray.cluster_resources()` to ensure the expected number of resources are taken up by the currently running script.
> * Add the name of the new YAML file to `test_sample_rayjob_yamls.py` so that it's tested in CI.
> 
> What do you think?



> Looks good to me, but there's a typo. To catch this we would have needed an integration test with Ray. Can you add one? Here's one idea.
> 
> * Add a new sample YAML file which has the job specify some cpus, gpus, and resources.  I think there should be some way to also specify these logical resources in the RayCluster spec for the job (since we don't have physical GPUs, we don't want to autodetect the number of GPUs, but we can just define that the cluster has 4 GPUs or something.)
> * In the entrypoint script, use `ray.available_resources()` and `ray.cluster_resources()` to ensure the expected number of resources are taken up by the currently running script.
> * Add the name of the new YAML file to `test_sample_rayjob_yamls.py` so that it's tested in CI.
> 
> What do you think?

Yes! That sounds great, Let me try adding that test!

**Comment by architkulkarni** [comment]
Great, let me know if you have any questions!

**Comment by shubhscoder** [comment]
> "don't

Actually what do you mean by this: ". I think there should be some way to also specify these logical resources in the RayCluster spec for the job (since we don't have physical GPUs, we don't want to autodetect the number of GPUs, but we can just define that the cluster has 4 GPUs or something.)
"

Following is my understanding:

1. RayClusterSpec specifies the desired state of the cluster. I see we can specify ResourceRequirements like CPU / memory and maybe even GPUs, etc.
2. Then kubernetes would try to map the pods of the ray cluster to nodes which have those resources?
3. So If we specify say the resources something like this:
```
      resources:
        limits:
          cpu: "1"     # Limit to 1 CPU core
          nvidia.com/gpu: 1  # Limit to 1 GPU
        requests:
          cpu: "200m"  # Request 200m CPU (0.2 CPU cores)
          nvidia.com/gpu: 1  # Request 1 GPU
```
In that case, the infra on which our tests run would need to have a GPU (which I assume we don't have). Then won't the cluster creation remain in a pending state? I think there is some gap in my understanding! Would be very helpful if you could clarify this! Thanks for your help!

**Comment by architkulkarni** [comment]
Yeah exactly, we don't have physical GPUs, but from Ray's perspective the resources are logical, not physical. So we can just tell Ray that a certain node has 4 GPUs even if it doesn't, and it will schedule tasks and actors (and in this case, the entrypoint script) accordingly.  https://docs.ray.io/en/latest/ray-core/scheduling/resources.html#specifying-node-resources (you can see the "KubeRay" tab)

**Comment by shubhscoder** [comment]
@architkulkarni I have pushed the tests. However, I was facing one issue while running the tests. The rayjob used to start and fail as the ray cluster was not ready (Atleast that is what I interepreted from the logs). However, the next attempt (a resubmission I guess) used to pass, because maybe the cluster was up and running by this time. I think the intended behavior is that the cluster comes up, it is ready and then we start/submit the job. 

Do you think this is a bug?
Also one problem is that the behavior is not consistently reproducible. I was seeing this behavior day before yesterday and yesterday, but I deleted my kind cluster and ran the submission again, and this time the behavior was as expected. So, unfortunately I don't have the exact way to reproduce this issue.

**Comment by architkulkarni** [comment]
@shubhscoder Ah thanks, that sounds like a bug. If you see it again or if you have the logs from last time would you mind filing an issue?  Thanks!

**Comment by architkulkarni** [comment]
https://github.com/ray-project/kuberay/actions/runs/5986738858/job/16284203893?pr=1359#step:5:10

Can you try this https://github.com/ray-project/kuberay/blob/master/ray-operator/DEVELOPMENT.md#consistency-check and see if the error goes away?

**Comment by shubhscoder** [comment]
@architkulkarni I tried running make sync locally and I keep getting this error:

test -s /vagrant/kuberay/ray-operator/bin/controller-gen || GOBIN=/vagrant/kuberay/ray-operator/bin/controller-gen/.. go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.6.0
/vagrant/kuberay/ray-operator/bin/controller-gen "crd:maxDescLen=100,trivialVersions=true,preserveUnknownFields=false,generateEmbeddedObjectMeta=true,allowDangerousTypes=true" rbac:roleName=kuberay-operator webhook paths="./..." output:crd:artifacts:config=config/crd/bases
test -s /vagrant/kuberay/ray-operator/bin/kustomize ||  GOBIN=/vagrant/kuberay/ray-operator/bin/kustomize/.. go install sigs.k8s.io/kustomize/kustomize/v3@v3.10.0
go: sigs.k8s.io/kustomize/kustomize/v3@v3.10.0 (in sigs.k8s.io/kustomize/kustomize/v3@v3.10.0):
        The go.mod file for the module providing named packages contains one or
        more exclude directives. It must not contain directives that would cause
        it to be interpreted differently than if it were the main module.
Makefile:117: recipe for target 'kustomize' failed
make: *** [kustomize] Error 1

Looks like I am missing something

**Comment by architkulkarni** [comment]
> @architkulkarni I tried running make sync locally and I keep getting this error:

Oh weird, I'm not super familiar with this part... cc @kevin85421 in case you have any ideas

Worst case I can just check out your PR and run the command and push to your branch, if you have that enabled.

**Comment by architkulkarni** [comment]
Would you mind creating an issue for the `kustomize` failure?  

Also, another option is to "manually" sync the files, using https://github.com/ray-project/ray/pull/38857 as an example

**Review by architkulkarni — COMMENTED** [review]
The code and the test LGTM otherwise!

**Comment by shubhscoder** [comment]
@architkulkarni Sure I can file a bug for Kustomize. This looks related to https://github.com/kubernetes-sigs/kustomize/issues/3618 

See specifically this comment: where the user tried to install Kustomize 4.0.1 and got the exact same error that I am getting. 
According to this comment the bug was fixed and the user was able to install 4.5.2 https://github.com/kubernetes-sigs/kustomize/issues/3618#issuecomment-1054010950

My guess is that this issue has started surfacing after upgrading to go 1.19, and maybe users who installed Kustomize with older versions of go did not face this issue and their old installations still seems to be working. (Just a guess). 

I got around it for now by changing the Kustomize version to 4.5.2 locally. However, I will still create an issue as I think this might affect other users and maybe we should be upgrading to 4.5.2 or higher anyways

**Review by architkulkarni — APPROVED** [review]

**Comment by shubhscoder** [comment]
@architkulkarni Thanks for all the help on this one! Also, apologies for so much too and fro on this fairly simple issue! I will try my best to make the future code changes more concise and tested.

**Comment by architkulkarni** [comment]
Not at all, I think it's normal. Thanks for the contribution!

**Comment by architkulkarni** [comment]
> @architkulkarni I have pushed the tests. However, I was facing one issue while running the tests. The rayjob used to start and fail as the ray cluster was not ready (Atleast that is what I interepreted from the logs). However, the next attempt (a resubmission I guess) used to pass, because maybe the cluster was up and running by this time. I think the intended behavior is that the cluster comes up, it is ready and then we start/submit the job.
> 
> Do you think this is a bug? Also one problem is that the behavior is not consistently reproducible. I was seeing this behavior day before yesterday and yesterday, but I re-created my kind cluster and ran the submission again, and this time the behavior was as expected. So, unfortunately I don't have the exact way to reproduce this issue.

Saw it a couple times, filed an issue here https://github.com/ray-project/kuberay/issues/1381

**Pull request merged**